### PR TITLE
Support selective inlining in function inliner

### PR DESCRIPTION
**Pull request by gramalingam**
### Description
This extends the function-inliner to support selective function inlining. The caller can indicate which functions are to be inlined.

Consider an example where the main-graph calls functions f1 and f2. Suppose further that f2 calls f1. If we ask for f1 to be inlined, but not f2, then the call to f1 in both main and f2 will be inlined. But f2 will not be inlined into main.

This introduces one significant change to the inlining logic: now, the call-site to be inlined may be one inside a FunctionProto (instead of a GraphProto). This has some implications, namely, full-context-information (types of input parameters or values of attribute parameters)_may not be known at inlining time. The inlining is still doable. However, more advanced features like version-conversion, when inlining, cannot be supported.

### CODE REVIEW

File: onnx/cpp2py_export.cc [hunk 1]
```diff
@@ -573,6 +573,21 @@ PYBIND11_MODULE(onnx_cpp2py_export, onnx_cpp2py_export) {
...
+      [](const py::bytes& bytes, std::vector<std::pair<std::string, std::string>> function_ids, bool exclude) {
+        ModelProto model{};
+        ParseProtoFromPyBytes(&model, bytes);
+        auto function_id_set = inliner::FunctionIdSet::Create(std::move(function_ids), exclude);
+        inliner::InlineSelectedFunctions(model, *function_id_set);
+        std::string out;
+        model.SerializeToString(&out);
+        return py::bytes(out);
+      });
+
```
Comment (justinchuby) [inline onnx/cpp2py_export.cc:590]:
```suggestion
  inliner.def(
      "inline_selected_functions",
      [](const py::bytes& bytes, std::vector<std::pair<std::string, std::string>> function_ids, bool exclude) {
        ModelProto model{};
        ParseProtoFromPyBytes(&model, bytes);
        auto function_id_set = inliner::FunctionIdSet::Create(std::move(function_ids), exclude);
        inliner::InlineSelectedFunctions(model, *function_id_set);
        std::string out;
        model.SerializeToString(&out);
        return py::bytes(out);
      },
      "Inlines all functions specified in function_ids, unless "
      "exclude is true, in which case it inlines all functions except those specified in "
      "function_ids."
      );

```

I would move the doc into the docstring

Comment (justinchuby) [reply onnx/cpp2py_export.cc:590]: NVM

File: onnx/inliner/inliner.cc [hunk 2]
```diff
@@ -67,6 +67,22 @@ struct OpsetMap : public OpsetMapBase {
     }
     return result;
   }
+
+  bool Add(const google::protobuf::RepeatedPtrField<OperatorSetIdProto>& list) {
```
[the remark this answers is not in the document]
Comment (gramalingam) [reply onnx/inliner/inliner.cc:88]: Done

**Review by justinchuby — APPROVED** [review]

File: onnx/onnx_cpp2py_export/inliner.pyi [hunk 3]
```diff
@@ -2,9 +2,20 @@
...
+from __future__ import annotations
+
 def inline_local_functions(model: bytes, convert_version: bool) -> bytes:
     """
     Inlines calls to model-local function in input model and returns it.
     Both input and output are serialized ModelProtos.
     """
     ...
+
+def inline_selected_functions(model: bytes, function_ids: list[tuple[str,str]], exclude: bool) -> bytes:
```
Comment (justinchuby) [inline onnx/onnx_cpp2py_export/inliner.pyi:14]: Should we give default to exclude?

File: onnx/test/inliner_test.py [hunk 4]
```diff
@@ -35,6 +35,78 @@ def test_basic(self):
...
+                y = local.square(double)
+            }
+
+            <opset_import: [ "" : 17 ], domain: "local">
+            square (x) => (y) {
+                y = Mul (x, x)
+            }
+        """
+        )
+        inlined = inliner.inline_selected_functions(model, [("local", "square")], False)
```
Comment (justinchuby) [inline onnx/test/inliner_test.py:60]:
```suggestion
        inlined = inliner.inline_selected_functions(model, [("local", "square")], exclude=False)
```

for clarity

File: onnx/test/inliner_test.py [hunk 5]
```diff
@@ -35,6 +35,78 @@ def test_basic(self):
...
+            }
+
+            <opset_import: [ "" : 17 ], domain: "local">
+            square (x) => (y) {
+                y = Mul (x, x)
+            }
+        """
+        )
+        inlined = inliner.inline_selected_functions(
+            model, [("local", "double_and_square")], True
```
Comment (justinchuby) [inline onnx/test/inliner_test.py:96]:
```suggestion
            model, [("local", "double_and_square")], exclude=True
```

File: onnx/cpp2py_export.cc [hunk 6]
```diff
@@ -573,6 +573,21 @@ PYBIND11_MODULE(onnx_cpp2py_export, onnx_cpp2py_export) {
     return py::bytes(out);
   });
 
+  // inline_selected_functions: Inlines all functions specified in function_ids, unless
+  // exclude is true, in which case it inlines all functions except those specified in
+  // function_ids.
+  inliner.def(
+      "inline_selected_functions",
```
Comment (xadupre) [inline onnx/cpp2py_export.cc:580]: The documentation should include the parameter definition. *exclude* is not intuitive.

File: onnx/test/cpp/inliner_test.cc [hunk 7]
```diff
@@ -28,13 +28,15 @@ static void InlineFunctions(ModelProto& model, const char* input) {
   shape_inference::InferShapes(model);
 
   // std::cout << "Before inlining:\n" << ProtoToString(model) << "\n";
-  inliner::InlineLocalFunctions(model, true);
+  if (to_inline != nullptr)
```
Comment (xadupre) [inline onnx/test/cpp/inliner_test.cc:31]: use `to_inline != nullptr ? *to_inline : true`

**Pull request merged**